### ISSUE TITLE
I have a code question about restyle-encoder/training/coach_restyle_psp.py

### ISSUE THREAD

**Issue by SeungJunnn** [yuval-alaluf/restyle-encoder#33]
Hello. Thank you for releasing an awesome code and a paper for StyleGAN Encoder. I read Restyle paper very interestingly.

I have a question about your codes, at 107-th line of restyle-encoder/training/coach_restyle_psp.py

```
for iter in range(self.opts.n_iters_per_batch):
	if iter == 0:
		avg_image_for_batch = self.avg_image.unsqueeze(0).repeat(x.shape[0], 1, 1, 1)
		x_input = torch.cat([x, avg_image_for_batch], dim=1)
		y_hat, latent = self.net.forward(x_input, latent=None, return_latents=True)
	else:
		y_hat_clone = y_hat.clone().detach().requires_grad_(True)
		latent_clone = latent.clone().detach().requires_grad_(True)
		x_input = torch.cat([x, y_hat_clone], dim=1)
		y_hat, latent = self.net.forward(x_input, latent=latent_clone, return_latents=True)

	if self.opts.dataset_type == "cars_encode":
		y_hat = y_hat[:, :, 32:224, :]

	loss, loss_dict, id_logs = self.calc_loss(x, y, y_hat, latent)
	loss.backward()
```

In here, I am curious why we do loss.backward() at every iteration step. Isn't it enough only backward at the end of the last iteration(in the given args, at the 5-th iteration)?

Thank you for your attention, and interesting works again!\

**Comment by yuval-alaluf** [comment]
Good question! 
You could perform the `loss.backward()` only at the final step. However, if you perform the backward at every step, you allow the encoder to learn also from the intermediate outputs rather than only the final output. I like to think of this as "pushing" the encoder gradually to the final output rather than forcing it from learning only from the final result (which is the case in standard single-pass encoders). 
I hope that answers your question :)
